### PR TITLE
kelly/reset_call_reset_put

### PR DESCRIPTION
**Pull request by kelly-binary**

### CODE REVIEW

File: src/javascript/app/pages/trade/charts/highchart.js [hunk 1]
```diff
@@ -364,14 +369,22 @@ const Highchart = (() => {
...
     const drawBarrier = () => {
         if (chart.yAxis[0].plotLinesAndBands.length === 0) {
-            const {contract_type, barrier, high_barrier, low_barrier} = contract;
+            const { barrier, contract_type, entry_spot, high_barrier, low_barrier } = contract;
             if (barrier) {
                 prev_barriers[0] = barrier; // Batman like the kids who "Cache".
                 if (Lookback.isLookback(contract_type)) {
                     const label = Lookback.getBarrierLabel(contract_type);
-                    addPlotLine({ id: 'barrier',      value: barrier * 1,      label: localize(`${label} ([_1])`, [addComma(barrier)]),           dashStyle: 'Dot' }, 'y');
+                    addPlotLine({ id: 'barrier',       value: barrier * 1,            label: localize(`${label} ([_1])`,  [addComma(barrier)]),          dashStyle: 'Dot'   }, 'y');
```
Comment (negarn) [inline src/javascript/app/pages/trade/charts/highchart.js]: if trying to align it with the lines below, `dashStyle` looks out of alignment :)

File: src/javascript/app/pages/trade/charts/highchart.js [hunk 2]
```diff
@@ -364,14 +369,22 @@ const Highchart = (() => {
...
+                    addPlotLine({ id: 'barrier',       value: entry_spot * 1,         label: localize('Barrier ([_1])', [addComma(entry_spot)]),    dashStyle: 'Dot'   }, 'y');
+                    addPlotLine({ id: 'reset_barrier', value: barrier * 1,            label: localize('Reset Barrier ([_1])', [addComma(barrier)]), dashStyle: 'Solid' }, 'y');
+                    drawLineX({
+                        value: (Reset.ticks.filter((tick) => (tick.y === barrier * 1))[0].x) / 1000,
+                        label: localize('Reset Time'),
+                        color: '#000',
+                    });
                 } else {
-                    addPlotLine({ id: 'barrier',      value: barrier * 1,      label: localize('Barrier ([_1])', [addComma(barrier)]),           dashStyle: 'Dot' }, 'y');
+                    addPlotLine({ id: 'barrier',       value: barrier * 1, label: localize('Barrier ([_1])', [addComma(barrier)]), dashStyle: 'Dot' },   'y');
```
Comment (negarn) [inline src/javascript/app/pages/trade/charts/highchart.js]: this whole line's alignment is weird :D

File: src/javascript/app/pages/trade/reset.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
+const CommonFunctions    = require('../../../_common/common_functions');
+const localize           = require('../../../_common/localize').localize;
```
Comment (negarn) [inline src/javascript/app/pages/trade/reset.js]: alignment of `=` :)

File: src/javascript/app/pages/trade/reset.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        CommonFunctions.getElementById('reset_time')
+            .html(localize('The reset time is [_1]', [reset_time_str]))
+            .setAttribute('style', '');
+    };
+
+    const hideResetTime = () => { CommonFunctions.getElementById('reset_time').style.display = 'none'; };
+
+    const isReset = (contract_type) => /^(RESETCALL|RESETPUT|reset)$/.test(contract_type);
```
Comment (negarn) [inline src/javascript/app/pages/trade/reset.js]: I think it's better to use lower case regex with `i` option, and isn't `reset` already handling `resetcall` and `resetput` checks too?

File: src/javascript/app/pages/trade/tick_trade.js [hunk 5]
```diff
@@ -86,7 +91,7 @@ const TickDisplay = (() => {
                 label: 'Exit Spot',
                 id   : 'exit_tick',
             };
-        } else if (contract_category.match('callput')) {
+        } else if (contract_category.match('callput') || contract_category.match('reset')) {
```
Comment (negarn) [inline src/javascript/app/pages/trade/tick_trade.js]: can change to regex check since checking two things now

File: src/javascript/app/pages/trade/tick_trade.js [hunk 6]
```diff
@@ -358,6 +367,55 @@ const TickDisplay = (() => {
...
+        const entry_barrier    = entry_spot * 1;
+        const reset_barrier    = r_barrier || absolute_barrier * 1;
+
+        if (entry_barrier !== reset_barrier) {
+            removePlotLine('tick-barrier', 'y');
+
+            chart.yAxis[0].addPlotLine({
+                id    : 'tick-reset-barrier',
+                value : reset_barrier,
+                label : { text: `Reset Barrier (${addComma(reset_barrier)})`, align: 'center' },
```
Comment (negarn) [inline src/javascript/app/pages/trade/tick_trade.js]: missing localize? same on below lines for label

Comment (negarn) [reply src/javascript/app/pages/trade/tick_trade.js]: suggestion: change to `localize('Reset Barrier ([_1])', addComma(reset_barrier))` so it matches the generate-static-data text

File: src/javascript/app/pages/user/view_popup/view_popup.js [hunk 7]
```diff
@@ -250,6 +253,10 @@ const ViewPopup = (() => {
             $container.find('#errMsg').setVisibility(0);
         }
 
+        const { barrier, contract_type, entry_spot } = contract;
+        if (Reset.isReset(contract_type) && (entry_spot !== barrier)) {
```
Comment (negarn) [inline src/javascript/app/pages/user/view_popup/view_popup.js]: do we need a `+` to make sure entry spot and barrier are both integers? 😅 forever suspicious of API formats

File: src/javascript/app/pages/trade/purchase.js [hunk 8]
```diff
@@ -49,7 +49,7 @@ const Purchase = (() => {
         const button             = CommonFunctions.getElementById('contract_purchase_button');
 
         const error      = details.error;
-        const show_chart = !error && passthrough.duration <= 10 && passthrough.duration_unit === 't' && /^(risefall|higherlower|asian|touchnotouch)$/.test(sessionStorage.formname);
+        const show_chart = !error && passthrough.duration <= 10 && passthrough.duration_unit === 't' && /^(risefall|higherlower|asian|touchnotouch|reset)$/.test(sessionStorage.formname);
```
Comment (negarn) [inline src/javascript/app/pages/trade/purchase.js]: I think at this point it's easier to say which contract types we don't want to show chart for. probably it's only digits?

File: src/javascript/app/pages/trade/tick_trade.js [hunk 9]
```diff
@@ -375,7 +390,47 @@ const TickDisplay = (() => {
...
+    const removePlotLine = (id, type = 'y') => {
+        if (!chart) return;
+        chart[(`${type}Axis`)][0].removePlotLine(id);
+    };
+
+    const plotResetSpot = (r_barrier) => {
+        if (reset_spot_plotted || !chart || !entry_spot) return;
+
+        const entry_barrier    = entry_spot * 1;
+        const reset_barrier    = (r_barrier || absolute_barrier) * 1;
```
Comment (negarn) [inline src/javascript/app/pages/trade/tick_trade.js]: `=` alignment has extra space

Comment (negarn) [reply src/javascript/app/pages/trade/tick_trade.js]: `+entry_spot` etc instead of ` * 1 `?

File: src/javascript/app/pages/trade/purchase.js [hunk 10]
```diff
@@ -49,7 +49,7 @@ const Purchase = (() => {
         const button             = CommonFunctions.getElementById('contract_purchase_button');
 
         const error      = details.error;
-        const show_chart = !error && passthrough.duration <= 10 && passthrough.duration_unit === 't' && /^(risefall|higherlower|asian|touchnotouch|reset)$/.test(sessionStorage.formname);
+        const show_chart = !error && passthrough.duration <= 10 && passthrough.duration_unit === 't' && !/^(digits)$/.test(sessionStorage.formname);
```
Comment (negarn) [inline src/javascript/app/pages/trade/purchase.js]: why not `sessionStorage.formname !== 'digits'`? up to you, in case we think in future we need more exceptions for chart

File: src/javascript/app/pages/trade/purchase.js [hunk 11]
```diff
@@ -49,7 +49,7 @@ const Purchase = (() => {
         const button             = CommonFunctions.getElementById('contract_purchase_button');
 
         const error      = details.error;
-        const show_chart = !error && passthrough.duration <= 10 && passthrough.duration_unit === 't' && /^(risefall|higherlower|asian|touchnotouch)$/.test(sessionStorage.formname);
+        const show_chart = !error && passthrough.duration <= 10 && passthrough.duration_unit === 't' && !/^(digits)$/.test(sessionStorage.formname);
```
Comment (negarn) [inline src/javascript/app/pages/trade/purchase.js]: I believe the formname for digits are `overunder|evenodd|matchdiff`, please test it, I don't think it works with `digits`

File: src/javascript/app/pages/trade/barriers.js [hunk 12]
```diff
@@ -21,7 +21,7 @@ const Barriers = (() => {
         const barriers  = Contract.barriers()[Defaults.get('underlying')];
         const form_name = Contract.form();
 
-        if (barriers && form_name && Defaults.get('formname') !== 'risefall') {
+        if (barriers && form_name && !/risefall|reset/i.test(Defaults.get('formname'))) {
```
Comment (negarn) [inline src/javascript/app/pages/trade/barriers.js]: probably we can ask API not to return barrier in contracts_for for reset if it's not needed for calling proposal.

File: scripts/generate-static-data.js [hunk 13]
```diff
@@ -412,6 +416,8 @@ const texts = [
     'High Barrier',
     'High Barrier ([_1])',
     'Barrier ([_1])',
+    'Reset Barrier',
+    'Reset Barrier ([_1])',
```
Comment (negarn) [inline scripts/generate-static-data.js]: how about instead of an extra text for `'Reset Barrier ([_1])'`, in js file we only localize Reset Barrier, then append the ` (...)` to the end?

File: src/javascript/app/pages/trade/barriers.js (hunk 12 above)
Comment (negarn) [reply src/javascript/app/pages/trade/barriers.js]: can you add a TODO comment to remove reset from here when API stops sending it so we don't forget?

File: src/javascript/app/pages/trade/charts/highchart.js [hunk 14]
```diff
@@ -363,37 +369,52 @@ const Highchart = (() => {
...
+                    addPlotLine({ id: 'high_barrier',  value: high_barrier * 1, label: localize(`${high_label} ([_1])`, [addComma(high_barrier)]), dashStyle: 'Dot' }, 'y');
+                    addPlotLine({ id: 'low_barrier',   value: low_barrier * 1,  label: localize(`${low_label} ([_1])`, [addComma(low_barrier)]),   dashStyle: 'Dot', textBottom: true }, 'y');
                 } else {
-                    addPlotLine({ id: 'high_barrier', value: high_barrier * 1, label: localize('High Barrier ([_1])', [addComma(high_barrier)]), dashStyle: 'Dot' }, 'y');
-                    addPlotLine({ id: 'low_barrier',  value: low_barrier * 1,  label: localize('Low Barrier ([_1])', [addComma(low_barrier)]),   dashStyle: 'Dot', textBottom: true }, 'y');
+                    addPlotLine({ id: 'high_barrier',  value: high_barrier * 1, label: localize('High Barrier ([_1])', [addComma(high_barrier)]),  dashStyle: 'Dot' }, 'y');
+                    addPlotLine({ id: 'low_barrier',   value: low_barrier * 1,  label: localize('Low Barrier ([_1])', [addComma(low_barrier)]),    dashStyle: 'Dot', textBottom: true }, 'y');
```
Comment (negarn) [inline src/javascript/app/pages/trade/charts/highchart.js]: can we revert the changes in these four lines (405, 406, 408, 409) since it's all whitespace and not needed anymore?

File: src/javascript/app/pages/trade/charts/highchart.ui.js [hunk 15]
```diff
@@ -4,12 +4,14 @@ const localize = require('../../../../_common/localize').localize;
...
     let common_time_style,
         common_spot_style,
+        reset_time_style,
         txt,
         chart_options;
 
     const initLabels = () => {
         common_time_style = 'margin-bottom: 3px; margin-left: 10px; height: 0; width: 20px; border: 0; border-bottom: 2px; border-color: #e98024; display: inline-block;';
         common_spot_style = 'margin-left: 10px; display: inline-block; border-radius: 6px;';
+        reset_time_style  = 'margin-bottom: 3px; margin-left: 10px; height: 0; width: 20px; border: 0; border-bottom: 2px; border-color: #000; display: inline-block;';
```
Comment (negarn) [inline src/javascript/app/pages/trade/charts/highchart.ui.js]: it looks exactly like `common_time_style` except for `border-color`, perhaps we can remove `border-color` from `common_time_style` and use that and add the `border-color` accordingly to each element using it

File: src/javascript/app/pages/trade/charts/highchart.ui.js [hunk 16]
```diff
@@ -25,19 +27,21 @@ const HighchartUI = (() => {
...
     };
 
-    const setLabels = (chart_delayed) => {
+    const setLabels = (chart_delayed, is_reset_contract) => {
         // display a guide for clients to know how we are marking entry and exit spots
         txt = (chart_delayed ? getLabels('delay') : '') +
             getLabels('start_time') +
             (history ? getLabels('entry_spot') + getLabels('exit_spot') : '') +
-            getLabels('end_time');
+            getLabels('end_time') + (is_reset_contract ? getLabels('reset_time') : '');
```
Comment (negarn) [inline src/javascript/app/pages/trade/charts/highchart.ui.js]: shouldn't reset time be in the middle of the legend? right now it's pretty chronological, if reset time occurs in the middle of entry spot and exit spot then we should put it there

File: src/javascript/app/pages/trade/purchase.js [hunk 17]
```diff
@@ -49,7 +49,7 @@ const Purchase = (() => {
         const button             = CommonFunctions.getElementById('contract_purchase_button');
 
         const error      = details.error;
-        const show_chart = !error && passthrough.duration <= 10 && passthrough.duration_unit === 't' && /^(risefall|higherlower|asian|touchnotouch)$/.test(sessionStorage.formname);
+        const show_chart = !error && passthrough.duration <= 10 && passthrough.duration_unit === 't' && !/^(overunder|evenodd|matchdiff)$/.test(sessionStorage.formname);
```
Comment (negarn) [inline src/javascript/app/pages/trade/purchase.js]: I think it's even better to use `Contract.form() === 'digits'`, similar to line 113, then no need to check all three separately

File: src/javascript/app/pages/trade/tick_trade.js [hunk 18]
```diff
@@ -368,7 +391,51 @@ const TickDisplay = (() => {
...
+
+    const removePlotLine = (id, type = 'y') => {
+        if (!chart) return;
+        chart[(`${type}Axis`)][0].removePlotLine(id);
+    };
+
+    const plotResetSpot = (r_barrier) => {
+        if (reset_spot_plotted || !chart) return;
+
+        const is_RESETCALL  = contract.contract_type === 'RESETCALL';
```
Comment (negarn) [inline src/javascript/app/pages/trade/tick_trade.js]: can we keep lowercase for variable names?

File: src/javascript/app/pages/trade/charts/highchart.ui.js (hunk 16 above)
Comment (kellybinary) [reply src/javascript/app/pages/trade/charts/highchart.ui.js]: sure, that's a better idea

File: scripts/generate-static-data.js (hunk 13 above)
Comment (kellybinary) [reply scripts/generate-static-data.js]: sure, should we change this for the Barrier, High Barrier, Low Barrier labels as well?

Comment (negarn) [reply scripts/generate-static-data.js]: up to you, since they're already translated

File: src/javascript/app/pages/trade/tick_trade.js [hunk 19]
```diff
@@ -231,7 +231,7 @@ const TickDisplay = (() => {
                 value: calc_barrier,
                 color: 'green',
                 label: {
-                    text : `Average (${addComma(calc_barrier)})`,
+                    text : `${localize('Average')} (${addComma(calc_barrier)})`,
```
Comment (negarn) [inline src/javascript/app/pages/trade/tick_trade.js]: I think we don't have 'Average' in generate static data, I had to add it in my other PR. also, do we have any more `Barrier ([_1])` (high barrier etc variations as well) anywhere? if no, we can remove them from generate static data

Comment (kellybinary) [reply src/javascript/app/pages/trade/tick_trade.js]: you already added it in your hi_lo PR: https://github.com/binary-com/binary-static/pull/3845/files#diff-4a55e57599e7d49fa3a6bdfedc88160bR241 :) 
and yeah, i missed out some `Barrier ([_1]) ` strings, will submit a fix here

Comment (negarn) [reply src/javascript/app/pages/trade/tick_trade.js]: yeah, but I don't know if my PR will get merged before yours or not :(

Comment (kellybinary) [reply src/javascript/app/pages/trade/tick_trade.js]: it's okay, we still need to localize `Average`, no? :P

**Comment by jy-binary** [comment]
conflicts

**Comment by negarn** [comment]
conflicts :)

**Comment by jy-binary** [comment]
conflicts
